<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://cantinilab.github.io/GRnnData/utils/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>utils - grnndata</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "utils";
        var mkdocs_page_input_path = "utils.md";
        var mkdocs_page_url = "/GRnnData/utils/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> grnndata
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../notebooks/example/">example notebook</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">documentation</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">utils</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#grnndata.utils">utils</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.TF">TF</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.mTF">mTF</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.compute_cluster">compute_cluster</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.compute_connectivities">compute_connectivities</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.enrichment">enrichment</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.fileToList">fileToList</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.get_centrality">get_centrality</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.metrics">metrics</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#grnndata.utils.metrics--small-worldness">small worldness</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#grnndata.utils.metrics--scale-freeness">scale freeness</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#grnndata.utils.plot_cluster">plot_cluster</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../grnanndata/">GRNAnnData</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">grnndata</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">documentation</li>
      <li class="breadcrumb-item active">utils</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="documentation-for-utils-module">Documentation for <code>utils</code> module</h1>


<div class="doc doc-object doc-module">



<h2 id="grnndata.utils" class="doc doc-heading">
            <code>grnndata.utils</code>


</h2>

    <div class="doc doc-contents first">



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="grnndata.utils.TF" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">TF</span> <span class="o">=</span> <span class="n">fileToList</span><span class="p">(</span><span class="n">file_dir</span> <span class="o">+</span> <span class="s1">&#39;/TF.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>List of transcription factors (TF) loaded from the file 'TF.txt'.
Each line in the file represents a transcription factor.</p>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="grnndata.utils.mTF" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mTF</span> <span class="o">=</span> <span class="n">fileToList</span><span class="p">(</span><span class="n">file_dir</span> <span class="o">+</span> <span class="s1">&#39;/mTF.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>List of modified transcription factors (mTF) loaded from the file 'mTF.txt'.
Each line in the file represents a modified transcription factor.</p>
    </div>

</div>



<div class="doc doc-object doc-function">


<h3 id="grnndata.utils.compute_cluster" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_cluster</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>compute_cluster uses the igraph library to perform clustering on the GRN.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
                  (<code><a class="autorefs autorefs-internal" title="grnndata.GRNAnnData" href="../grnanndata/#grnndata.GRNAnnData">GRNAnnData</a></code>)
              –
              <div class="doc-md-description">
                <p>The gene regulatory network to analyze.</p>
              </div>
            </li>
            <li>
              <b><code>resolution</code></b>
                  (<code>float</code>, default:
                      <code>1.5</code>
)
              –
              <div class="doc-md-description">
                <p>The resolution parameter for the clustering algorithm. Defaults to 1.5.</p>
              </div>
            </li>
            <li>
              <b><code>seed</code></b>
                  (<code>int</code>, default:
                      <code>42</code>
)
              –
              <div class="doc-md-description">
                <p>The random seed for reproducibility. Defaults to 42.</p>
              </div>
            </li>
            <li>
              <b><code>use</code></b>
                  (<code>str</code>, default:
                      <code>&#39;louvain&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The clustering algorithm to use ('louvain' or 'leiden'). Defaults to "louvain".</p>
              </div>
            </li>
            <li>
              <b><code>n_iterations</code></b>
                  (<code>int</code>, default:
                      <code>2</code>
)
              –
              <div class="doc-md-description">
                <p>The number of iterations for the clustering algorithm. Defaults to 2.</p>
              </div>
            </li>
            <li>
              <b><code>max_comm_size</code></b>
                  (<code>int</code>, default:
                      <code>0</code>
)
              –
              <div class="doc-md-description">
                <p>The maximum community size for the clustering algorithm. Defaults to 0.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>ValueError</code>
              –
              <div class="doc-md-description">
                <p>If connectivities are not found in the GRN object.</p>
              </div>
            </li>
            <li>
                  <code>ValueError</code>
              –
              <div class="doc-md-description">
                <p>If the clustering algorithm specified in 'use' is not 'louvain' or 'leiden'.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>GRNAnnData</code></b>              –
              <div class="doc-md-description">
                <p>The GRNAnnData object with the computed clusters added to the 'var' DataFrame.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>grnndata/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_cluster</span><span class="p">(</span>
    <span class="n">grn</span><span class="p">:</span> <span class="n">GRNAnnData</span><span class="p">,</span>
    <span class="n">resolution</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span>
    <span class="n">seed</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">42</span><span class="p">,</span>
    <span class="n">use</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;louvain&quot;</span><span class="p">,</span>
    <span class="n">n_iterations</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">max_comm_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    compute_cluster uses the igraph library to perform clustering on the GRN.</span>

<span class="sd">    Args:</span>
<span class="sd">        grn (GRNAnnData): The gene regulatory network to analyze.</span>
<span class="sd">        resolution (float, optional): The resolution parameter for the clustering algorithm. Defaults to 1.5.</span>
<span class="sd">        seed (int, optional): The random seed for reproducibility. Defaults to 42.</span>
<span class="sd">        use (str, optional): The clustering algorithm to use (&#39;louvain&#39; or &#39;leiden&#39;). Defaults to &quot;louvain&quot;.</span>
<span class="sd">        n_iterations (int, optional): The number of iterations for the clustering algorithm. Defaults to 2.</span>
<span class="sd">        max_comm_size (int, optional): The maximum community size for the clustering algorithm. Defaults to 0.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If connectivities are not found in the GRN object.</span>
<span class="sd">        ValueError: If the clustering algorithm specified in &#39;use&#39; is not &#39;louvain&#39; or &#39;leiden&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        GRNAnnData: The GRNAnnData object with the computed clusters added to the &#39;var&#39; DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;connectivities&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Connectivities not found in the GRN object.&quot;</span><span class="p">)</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">_utils</span><span class="o">.</span><span class="n">get_igraph_from_adjacency</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;connectivities&quot;</span><span class="p">],</span> <span class="n">directed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="s2">&quot;weight&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">use</span> <span class="o">==</span> <span class="s2">&quot;louvain&quot;</span><span class="p">:</span>
        <span class="n">part</span> <span class="o">=</span> <span class="n">louvain</span><span class="o">.</span><span class="n">find_partition</span><span class="p">(</span>
            <span class="n">g</span><span class="p">,</span>
            <span class="n">louvain</span><span class="o">.</span><span class="n">RBConfigurationVertexPartition</span><span class="p">,</span>
            <span class="n">resolution_parameter</span><span class="o">=</span><span class="n">resolution</span><span class="p">,</span>
            <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">,</span>
            <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">use</span> <span class="o">==</span> <span class="s2">&quot;leiden&quot;</span><span class="p">:</span>
        <span class="n">part</span> <span class="o">=</span> <span class="n">leidenalg</span><span class="o">.</span><span class="n">find_partition</span><span class="p">(</span>
            <span class="n">g</span><span class="p">,</span>
            <span class="n">leidenalg</span><span class="o">.</span><span class="n">ModularityVertexPartition</span><span class="p">,</span>
            <span class="c1"># resolution_parameter=resolution,</span>
            <span class="n">n_iterations</span><span class="o">=</span><span class="n">n_iterations</span><span class="p">,</span>
            <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">,</span>
            <span class="n">max_comm_size</span><span class="o">=</span><span class="n">max_comm_size</span><span class="p">,</span>
            <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;use must be one of &#39;louvain&#39; or &#39;leiden&#39;&quot;</span><span class="p">)</span>
    <span class="n">groups</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">part</span><span class="o">.</span><span class="n">membership</span><span class="p">)</span>
    <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;cluster_&quot;</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">resolution</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
        <span class="n">values</span><span class="o">=</span><span class="n">groups</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;U&quot;</span><span class="p">),</span>
        <span class="n">categories</span><span class="o">=</span><span class="n">natsorted</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">groups</span><span class="p">))),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">grn</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="grnndata.utils.compute_connectivities" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_connectivities</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>This function computes the connectivities of a given Gene Regulatory Network (GRN).</p>
<p>It uses the NetworkX library to convert the GRN into a graph and then computes the spanner of the graph.
The spanner of a graph is a subgraph that approximates the original graph in terms of distances between nodes.
The stretch parameter determines the maximum distance between nodes in the spanner compared to the original graph.
The computed connectivities are then stored in the GRN object.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network for which the connectivities are to be computed.</p>
              </div>
            </li>
            <li>
              <b><code>stretch</code></b>
              –
              <div class="doc-md-description">
                <p>The maximum distance between nodes in the spanner compared to the original graph. Default is 2.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
grn : The Gene Regulatory Network with the computed connectivities.</p>

            <details class="quote">
              <summary>Source code in <code>grnndata/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_connectivities</span><span class="p">(</span><span class="n">grn</span><span class="p">:</span> <span class="n">GRNAnnData</span><span class="p">,</span> <span class="n">topK</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GRNAnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function computes the connectivities of a given Gene Regulatory Network (GRN).</span>

<span class="sd">    It uses the NetworkX library to convert the GRN into a graph and then computes the spanner of the graph.</span>
<span class="sd">    The spanner of a graph is a subgraph that approximates the original graph in terms of distances between nodes.</span>
<span class="sd">    The stretch parameter determines the maximum distance between nodes in the spanner compared to the original graph.</span>
<span class="sd">    The computed connectivities are then stored in the GRN object.</span>

<span class="sd">    Args:</span>
<span class="sd">        grn : The Gene Regulatory Network for which the connectivities are to be computed.</span>
<span class="sd">        stretch : The maximum distance between nodes in the spanner compared to the original graph. Default is 2.</span>

<span class="sd">    Returns:</span>
<span class="sd">    grn : The Gene Regulatory Network with the computed connectivities.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">grn_mask</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span> <span class="o">&gt;=</span> <span class="n">row</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="n">topK</span><span class="p">)</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">connect</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">connect</span><span class="p">[</span><span class="o">~</span><span class="n">grn_mask</span><span class="o">.</span><span class="n">values</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;connectivities&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">connect</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">grn</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="grnndata.utils.enrichment" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">enrichment</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>This function performs enrichment analysis on a given gene regulatory network (grn).</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
                  (<code><a class="autorefs autorefs-internal" title="grnndata.GRNAnnData" href="../grnanndata/#grnndata.GRNAnnData">GRNAnnData</a></code>)
              –
              <div class="doc-md-description">
                <p>The gene regulatory network to analyze.</p>
              </div>
            </li>
            <li>
              <b><code>of</code></b>
                  (<code>str</code>, default:
                      <code>&#39;Targets&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The specific component of the grn to focus on.</p>
              </div>
            </li>
            <li>
              <b><code>top_k</code></b>
                  (<code>int</code>, default:
                      <code>30</code>
)
              –
              <div class="doc-md-description">
                <p>The number of top results to return. Defaults to 10.</p>
              </div>
            </li>
            <li>
              <b><code>doplot</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to generate a plot of the results. Defaults to False.</p>
              </div>
            </li>
            <li>
              <b><code>gene_sets</code></b>
                  (<code>list</code>, default:
                      <code>[&#39;KEGG_2016&#39;, &#39;ENCODE_TF_ChIP-seq_2014&#39;, &#39;GO_Molecular_Function_2015&#39;, {&#39;TFs&#39;: <a class="autorefs autorefs-internal" title="grnndata.utils.TF" href="#grnndata.utils.TF">TF</a>}, <span title="grnndata.utils.file_dir">file_dir</span> + &#39;/celltype.gmt&#39;, &#39;OMIM_Disease&#39;, &#39;WikiPathways_2016&#39;, &#39;GO_Cellular_Component_2015&#39;, &#39;GTEx_Tissue_Sample_Gene_Expression_Profiles_up&#39;, &#39;TargetScan_microRNA&#39;, &#39;Chromosome_Location&#39;, &#39;PPI_Hub_Proteins&#39;]</code>
)
              –
              <div class="doc-md-description">
                <p>The gene sets to use for the enrichment analysis. Defaults to the gene sets provided in the function.</p>
              </div>
            </li>
            <li>
              <b><code>min_size</code></b>
                  (<code>int</code>, default:
                      <code>3</code>
)
              –
              <div class="doc-md-description">
                <p>The minimum size of the gene sets to consider. Defaults to 3.</p>
              </div>
            </li>
            <li>
              <b><code>max_size</code></b>
                  (<code>int</code>, default:
                      <code>4000</code>
)
              –
              <div class="doc-md-description">
                <p>The maximum size of the gene sets to consider. Defaults to 4000.</p>
              </div>
            </li>
            <li>
              <b><code>permutation_num</code></b>
                  (<code>int</code>, default:
                      <code>1000</code>
)
              –
              <div class="doc-md-description">
                <p>The number of permutations to use for the enrichment analysis. Defaults to 1000.</p>
              </div>
            </li>
            <li>
              <b><code>**kwargs</code></b>
              –
              <div class="doc-md-description">
                <p>Additional keyword arguments to be passed to the gseapy.prerank function.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pandas.DataFrame: A DataFrame containing the results of the enrichment analysis.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>grnndata/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">enrichment</span><span class="p">(</span>
    <span class="n">grn</span><span class="p">:</span> <span class="n">GRNAnnData</span><span class="p">,</span>
    <span class="n">of</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Targets&quot;</span><span class="p">,</span>
    <span class="n">doplot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">top_k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span>
    <span class="n">gene_sets</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;KEGG_2016&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ENCODE_TF_ChIP-seq_2014&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GO_Molecular_Function_2015&quot;</span><span class="p">,</span>
        <span class="p">{</span><span class="s2">&quot;TFs&quot;</span><span class="p">:</span> <span class="n">TF</span><span class="p">},</span>
        <span class="n">file_dir</span> <span class="o">+</span> <span class="s2">&quot;/celltype.gmt&quot;</span><span class="p">,</span>
        <span class="s2">&quot;OMIM_Disease&quot;</span><span class="p">,</span>
        <span class="s2">&quot;WikiPathways_2016&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GO_Cellular_Component_2015&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GTEx_Tissue_Sample_Gene_Expression_Profiles_up&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TargetScan_microRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Chromosome_Location&quot;</span><span class="p">,</span>
        <span class="s2">&quot;PPI_Hub_Proteins&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">min_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">max_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4000</span><span class="p">,</span>
    <span class="n">permutation_num</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>  <span class="c1"># reduce number to speed up testing</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function performs enrichment analysis on a given gene regulatory network (grn).</span>

<span class="sd">    Args:</span>
<span class="sd">        grn (GRNAnnData): The gene regulatory network to analyze.</span>
<span class="sd">        of (str, optional): The specific component of the grn to focus on.</span>
<span class="sd">        top_k (int, optional): The number of top results to return. Defaults to 10.</span>
<span class="sd">        doplot (bool, optional): Whether to generate a plot of the results. Defaults to False.</span>
<span class="sd">        gene_sets (list, optional): The gene sets to use for the enrichment analysis. Defaults to the gene sets provided in the function.</span>
<span class="sd">        min_size (int, optional): The minimum size of the gene sets to consider. Defaults to 3.</span>
<span class="sd">        max_size (int, optional): The maximum size of the gene sets to consider. Defaults to 4000.</span>
<span class="sd">        permutation_num (int, optional): The number of permutations to use for the enrichment analysis. Defaults to 1000.</span>
<span class="sd">        **kwargs: Additional keyword arguments to be passed to the gseapy.prerank function.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: A DataFrame containing the results of the enrichment analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">of</span> <span class="o">==</span> <span class="s2">&quot;Targets&quot;</span><span class="p">:</span>
        <span class="n">rnk</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">of</span> <span class="o">==</span> <span class="s2">&quot;Regulators&quot;</span><span class="p">:</span>
        <span class="n">rnk</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">of</span> <span class="o">==</span> <span class="s2">&quot;Central&quot;</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">get_centrality</span><span class="p">(</span><span class="n">grn</span><span class="p">,</span> <span class="n">top_k_to_disp</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">nx</span><span class="o">.</span><span class="n">PowerIterationFailedConvergence</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;PowerIterationFailedConvergence&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="n">rnk</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;centrality&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;of must be one of &#39;Targets&#39;, &#39;Regulators&#39;, or &#39;Central&#39;&quot;</span><span class="p">)</span>
    <span class="n">rnk</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">rnk</span> <span class="o">=</span> <span class="n">rnk</span><span class="p">[</span><span class="n">rnk</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The DataFrame contains only the same values.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="c1"># run enrichment analysis</span>
    <span class="n">previous_level</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">root</span><span class="o">.</span><span class="n">manager</span><span class="o">.</span><span class="n">disable</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">disable</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">WARNING</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">pre_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">prerank</span><span class="p">(</span>
            <span class="n">rnk</span><span class="o">=</span><span class="n">rnk</span><span class="p">,</span>  <span class="c1"># or rnk = rnk,</span>
            <span class="n">gene_sets</span><span class="o">=</span><span class="n">gene_sets</span><span class="p">,</span>
            <span class="n">min_size</span><span class="o">=</span><span class="n">min_size</span><span class="p">,</span>
            <span class="n">max_size</span><span class="o">=</span><span class="n">max_size</span><span class="p">,</span>
            <span class="n">background</span><span class="o">=</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>  <span class="c1"># or &quot;hsapiens_gene_ensembl&quot;, or int, or text file  or a list of genes</span>
            <span class="n">permutation_num</span><span class="o">=</span><span class="n">permutation_num</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">disable</span><span class="p">(</span><span class="n">previous_level</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">LookupError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;raised a lookup error&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">val</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[(</span><span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">])</span>
    <span class="p">)</span>

    <span class="c1"># plot results</span>
    <span class="k">if</span> <span class="n">doplot</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">val</span><span class="o">.</span><span class="n">Term</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[:</span><span class="n">top_k</span><span class="p">])</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">dotplot</span><span class="p">(</span>
            <span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span>
                <span class="p">(</span><span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
            <span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
            <span class="n">column</span><span class="o">=</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">,</span>
            <span class="n">title</span><span class="o">=</span><span class="s2">&quot;enrichment of &quot;</span> <span class="o">+</span> <span class="n">of</span> <span class="o">+</span> <span class="s2">&quot; in the grn&quot;</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>  <span class="c1"># adjust dot size</span>
            <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
            <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span>
            <span class="n">show_ring</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">pre_res</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="grnndata.utils.fileToList" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fileToList</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>loads an input file with a\n b\n.. into a list [a,b,..]</p>

            <details class="quote">
              <summary>Source code in <code>grnndata/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fileToList</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">strconv</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">],</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    loads an input file with a\\n b\\n.. into a list [a,b,..]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">strconv</span><span class="p">(</span><span class="n">val</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="grnndata.utils.get_centrality" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_centrality</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>get_centrality uses the networkx library to calculate the centrality of each node in the GRN.</p>
<p>The centrality is added to the grn object as a new column in the var dataframe.
also prints the top K most central nodes in the GRN.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
                  (<code><a class="autorefs autorefs-internal" title="grnndata.GRNAnnData" href="../grnanndata/#grnndata.GRNAnnData">GRNAnnData</a></code>)
              –
              <div class="doc-md-description">
                <p>The gene regulatory network to analyze.</p>
              </div>
            </li>
            <li>
              <b><code>n_largest_to_consider</code></b>
                  (<code>int</code>, default:
                      <code>20</code>
)
              –
              <div class="doc-md-description">
                <p>The number of largest values to consider when constructing the graph. Defaults to 20.</p>
              </div>
            </li>
            <li>
              <b><code>top_k_to_disp</code></b>
                  (<code>int</code>, default:
                      <code>30</code>
)
              –
              <div class="doc-md-description">
                <p>The number of top results to return. Defaults to 30.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>list</code>
              –
              <div class="doc-md-description">
                <p>A list of the top K most central genes in the GRN (sorted by centrality</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>grnndata/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span>
<span class="normal">99</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_centrality</span><span class="p">(</span>
    <span class="n">grn</span><span class="p">:</span> <span class="n">GRNAnnData</span><span class="p">,</span> <span class="n">n_largest_to_consider</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">top_k_to_disp</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    get_centrality uses the networkx library to calculate the centrality of each node in the GRN.</span>

<span class="sd">    The centrality is added to the grn object as a new column in the var dataframe.</span>
<span class="sd">    also prints the top K most central nodes in the GRN.</span>

<span class="sd">    Args:</span>
<span class="sd">        grn (GRNAnnData): The gene regulatory network to analyze.</span>
<span class="sd">        n_largest_to_consider (int, optional): The number of largest values to consider when constructing the graph. Defaults to 20.</span>
<span class="sd">        top_k_to_disp (int, optional): The number of top results to return. Defaults to 30.</span>

<span class="sd">    Returns:</span>
<span class="sd">        (list): A list of the top K most central genes in the GRN (sorted by centrality</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">grn</span> <span class="o">=</span> <span class="n">compute_connectivities</span><span class="p">(</span><span class="n">grn</span><span class="p">,</span> <span class="n">n_largest_to_consider</span><span class="p">)</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">from_scipy_sparse_array</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;connectivities&quot;</span><span class="p">],</span> <span class="n">create_using</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">)</span>
    <span class="n">centrality</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">eigenvector_centrality</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;centrality&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">centrality</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">top_central_genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="p">[(</span><span class="n">gene</span><span class="p">,</span> <span class="n">centrality</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">centrality</span> <span class="ow">in</span> <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;centrality&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()],</span>
        <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
        <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)[:</span><span class="n">top_k_to_disp</span><span class="p">]</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Top central genes:&quot;</span><span class="p">,</span> <span class="n">top_central_genes</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">top_central_genes</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="grnndata.utils.metrics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">metrics</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>metrics This function computes the following metrics for a given Gene Regulatory Network (GRN):</p>
<h6 id="grnndata.utils.metrics--small-worldness">small worldness</h6>
<p>A small-world network is a type of mathematical graph in which most nodes are not neighbors of one another,
but most nodes can be reached from every other by a small number of hops or steps.
the sigma metric is a measure of small-worldness.
if sigma &gt; 1, the network is a small-world network.</p>
<h6 id="grnndata.utils.metrics--scale-freeness">scale freeness</h6>
<p>A scale-free network is a network whose degree distribution follows a power law,
at least asymptotically. The s metric is a measure of scale-freeness,
defined as ( S(G)={frac {s(G)}{s_{max }}} ), where ( s_{max } )
is the maximum value of s(H) for H in the set of all graphs with degree distribution.
A graph with small S(G) is "scale-rich," and a graph with S(G) close to 1 is "scale-free"</p>
<p>Therefore, a graph is more scale-free when its S(G) value is closer to 1.
The range of the s metric is between 0 and 1, where 0 indicates "scale-rich" and 1 indicates "scale-free"</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network for which the connectivities are to be computed.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>{is_connected: bool, scale_freeness: float}</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>grnndata/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">metrics</span><span class="p">(</span><span class="n">grn</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    metrics This function computes the following metrics for a given Gene Regulatory Network (GRN):</span>

<span class="sd">    ### small worldness</span>
<span class="sd">    A small-world network is a type of mathematical graph in which most nodes are not neighbors of one another,</span>
<span class="sd">    but most nodes can be reached from every other by a small number of hops or steps.</span>
<span class="sd">    the sigma metric is a measure of small-worldness.</span>
<span class="sd">    if sigma &gt; 1, the network is a small-world network.</span>

<span class="sd">    ### scale freeness</span>
<span class="sd">    A scale-free network is a network whose degree distribution follows a power law,</span>
<span class="sd">    at least asymptotically. The s metric is a measure of scale-freeness,</span>
<span class="sd">    defined as ( S(G)={frac {s(G)}{s_{max }}} ), where ( s_{max } )</span>
<span class="sd">    is the maximum value of s(H) for H in the set of all graphs with degree distribution.</span>
<span class="sd">    A graph with small S(G) is &quot;scale-rich,&quot; and a graph with S(G) close to 1 is &quot;scale-free&quot;</span>

<span class="sd">    Therefore, a graph is more scale-free when its S(G) value is closer to 1.</span>
<span class="sd">    The range of the s metric is between 0 and 1, where 0 indicates &quot;scale-rich&quot; and 1 indicates &quot;scale-free&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        grn : The Gene Regulatory Network for which the connectivities are to be computed.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict : {is_connected: bool, scale_freeness: float}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">from_numpy_array</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;GRN&quot;</span><span class="p">])</span>
    <span class="c1"># sw_sig = nx.sigma(G) # commented because too long</span>
    <span class="n">conn</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">is_connected</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    <span class="c1"># clust_coef = nx.average_clustering(G)</span>
    <span class="n">degree_sequence</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="p">[</span><span class="n">d</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">degree</span><span class="p">()],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>  <span class="c1"># used for degree distribution and powerlaw test</span>
    <span class="n">fit</span> <span class="o">=</span> <span class="n">powerlaw</span><span class="o">.</span><span class="n">Fit</span><span class="p">(</span><span class="n">degree_sequence</span><span class="p">)</span>
    <span class="n">R</span><span class="p">,</span> <span class="n">p</span> <span class="o">=</span> <span class="n">fit</span><span class="o">.</span><span class="n">distribution_compare</span><span class="p">(</span><span class="s2">&quot;power_law&quot;</span><span class="p">,</span> <span class="s2">&quot;exponential&quot;</span><span class="p">,</span> <span class="n">normalized_ratio</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">fit</span><span class="o">.</span><span class="n">plot_pdf</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">fit</span><span class="o">.</span><span class="n">distribution_compare</span><span class="p">(</span><span class="s2">&quot;power_law&quot;</span><span class="p">,</span> <span class="s2">&quot;lognormal&quot;</span><span class="p">)</span>
    <span class="n">fig4</span> <span class="o">=</span> <span class="n">fit</span><span class="o">.</span><span class="n">plot_ccdf</span><span class="p">(</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
    <span class="n">fit</span><span class="o">.</span><span class="n">power_law</span><span class="o">.</span><span class="n">plot_ccdf</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">fig4</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>  <span class="c1"># powerlaw</span>
    <span class="n">fit</span><span class="o">.</span><span class="n">lognormal</span><span class="o">.</span><span class="n">plot_ccdf</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">fig4</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>  <span class="c1"># lognormal</span>
    <span class="n">fit</span><span class="o">.</span><span class="n">stretched_exponential</span><span class="o">.</span><span class="n">plot_ccdf</span><span class="p">(</span>
        <span class="n">ax</span><span class="o">=</span><span class="n">fig4</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span>
    <span class="p">)</span>  <span class="c1"># stretched_exponential</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="c1"># &quot;small_worldness&quot;: sw_sig,</span>
        <span class="s2">&quot;is_connected&quot;</span><span class="p">:</span> <span class="n">conn</span><span class="p">,</span>
        <span class="c1">#  &quot;clustering&quot;: clust_coef,</span>
        <span class="s2">&quot;scale_freeness&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">R</span><span class="p">,</span> <span class="n">p</span><span class="p">],</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="grnndata.utils.plot_cluster" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cluster</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>This function plots the clusters of a given Gene Regulatory Network (GRN).</p>
<p>It first computes the connectivities of the GRN and then performs Louvain clustering on the transpose of the GRN.
The clusters are then visualized using UMAP.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network to be clustered and visualized.</p>
              </div>
            </li>
            <li>
              <b><code>color</code></b>
              –
              <div class="doc-md-description">
                <p>The color of the clusters. Default is "louvain".</p>
              </div>
            </li>
            <li>
              <b><code>min_dist</code></b>
              –
              <div class="doc-md-description">
                <p>The minimum distance between points in the UMAP. Default is 0.5.</p>
              </div>
            </li>
            <li>
              <b><code>spread</code></b>
              –
              <div class="doc-md-description">
                <p>The spread of the points in the UMAP. Default is 0.7.</p>
              </div>
            </li>
            <li>
              <b><code>stretch</code></b>
              –
              <div class="doc-md-description">
                <p>The maximum distance between nodes in the spanner compared to the original graph. Default is 2.</p>
              </div>
            </li>
            <li>
              <b><code>**kwargs</code></b>
              –
              <div class="doc-md-description">
                <p>Additional keyword arguments to be passed to the UMAP function.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>grnndata/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">plot_cluster</span><span class="p">(</span><span class="n">grn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;louvain&quot;</span><span class="p">],</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">spread</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">stretch</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function plots the clusters of a given Gene Regulatory Network (GRN).</span>

<span class="sd">    It first computes the connectivities of the GRN and then performs Louvain clustering on the transpose of the GRN.</span>
<span class="sd">    The clusters are then visualized using UMAP.</span>

<span class="sd">    Args:</span>
<span class="sd">        grn : The Gene Regulatory Network to be clustered and visualized.</span>
<span class="sd">        color : The color of the clusters. Default is &quot;louvain&quot;.</span>
<span class="sd">        min_dist : The minimum distance between points in the UMAP. Default is 0.5.</span>
<span class="sd">        spread : The spread of the points in the UMAP. Default is 0.7.</span>
<span class="sd">        stretch : The maximum distance between nodes in the spanner compared to the original graph. Default is 2.</span>
<span class="sd">        **kwargs : Additional keyword arguments to be passed to the UMAP function.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">grn</span> <span class="o">=</span> <span class="n">compute_connectivities</span><span class="p">(</span><span class="n">grn</span><span class="p">,</span> <span class="n">stretch</span><span class="o">=</span><span class="n">stretch</span><span class="p">)</span>
    <span class="n">subgrn</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">T</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">louvain</span><span class="p">(</span><span class="n">subgrn</span><span class="p">,</span> <span class="n">adjacency</span><span class="o">=</span><span class="n">subgrn</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s2">&quot;GRN&quot;</span><span class="p">])</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">subgrn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../notebooks/example/" class="btn btn-neutral float-left" title="example notebook"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../grnanndata/" class="btn btn-neutral float-right" title="GRNAnnData">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../notebooks/example/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../grnanndata/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
